module/zfs/zfs_vnops.c

```diff
@@ -3563,6 +3652,17 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 	}
 
 	/*
+	 * If we are using project inheritance, we only allow renames
+	 * into our tree when the project IDs are the same; otherwise
+	 * tree quota mechanism would be circumvented.
```

Comment:
I am not sure whether it is "requirement", but it is useful. In fact, some other filesystem, such as EXT4, has already support "tree quota" like that. Means if the ZFS_PROJINHERIT flag is set on the directory, then all its descendants will inherit its project ID.